wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -1597,10 +1590,18 @@ private class WasmExpressionBuilder private (
   private def genThis(t: IRTrees.This): IRTypes.Type = {
     genReadStorage(fctx.receiverStorage)
 
+    // The synthetic `This` may contains boxed primitive, instead of primitive types.
+    // However, the checker should ensure that the `This` is never boxed in Scala.js linker frontend.
+    // As a workaround, we unbox the `This` here.
+    val fixedTpe = t.tpe match {
```

Comment:
Any better explanation?

Reply:
Perhaps as follows:

> Workaround wrong `t.tpe` for `This` nodes inside reflective proxies.
> In a hijacked class, `This` nodes are supposed to be typed as the corresponding primitive type.
> However, the Scala.js linker frontend synthesizes reflective proxies that contain `This` nodes typed as the hijacked class' `ClassType` instead.
> This is bad for us because it means `genAdapt` fails to box the primitives when required.
> We work around this issue here by re-computing the correct type of `This` nodes.